### PR TITLE
Collectable shared worker executor

### PR DESCRIPTION
**Pull request by vietj**

### CODE REVIEW

**Review by JojOatXGME — COMMENTED** [review]
Looks fine as far as I can tell. I am not sure if you are still interesting in my review, since you have already merged your changes. Anyway, here are my uninformed comments. (I actually started the review yesterday but was interrupted by a phone call.)

File: src/main/java/io/vertx/core/net/impl/NetClientImpl.java [hunk 1]
```diff
@@ -293,14 +293,5 @@ private void failed(ContextInternal context, Channel ch, Throwable th, Promise<N
...
     context.emit(th, connectHandler::tryFail);
   }
-
-  @Override
-  protected void finalize() throws Throwable {
-    // Make sure this gets cleaned up if there are no more references to it
-    // so as not to leave connections and resources dangling until the system is shutdown
-    // which could make the JVM run out of file handles.
-    closeFuture.close();
-    super.finalize();
```
Comment (JojOatXGME) [inline src/main/java/io/vertx/core/net/impl/NetClientImpl.java:303]: I noticed you removed the finalizers from `NetClientImpl`, `TCPServerBase`, and `SharedHTTPClient`. Do all these classes have there own `CloseFuture`?

Comment (vietj) [reply src/main/java/io/vertx/core/net/impl/NetClientImpl.java:303]: yes they do and the new collectable close future replaces those finalizers

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 2]
```diff
@@ -1048,9 +1048,11 @@ public synchronized WorkerExecutorImpl createSharedWorkerExecutor(String name, i
...
   public synchronized WorkerExecutorImpl createSharedWorkerExecutor(String name, int poolSize, long maxExecuteTime, TimeUnit maxExecuteTimeUnit) {
     SharedWorkerPool sharedWorkerPool = createSharedWorkerPool(name, poolSize, maxExecuteTime, maxExecuteTimeUnit);
-    CloseFuture fut = resolveCloseFuture();
-    WorkerExecutorImpl namedExec = new WorkerExecutorImpl(this, closeFuture, sharedWorkerPool);
-    fut.add(namedExec);
+    CloseFuture parentCf = resolveCloseFuture();
+    CloseFuture execCf = new CloseFuture();
+    parentCf.add(execCf);
+    WorkerExecutorImpl namedExec = new WorkerExecutorImpl(this, execCf, sharedWorkerPool);
+    execCf.add(namedExec);
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/core/impl/VertxImpl.java:1060]: it's not on GC, it references a shared thread pool that has no executor refs to it

File: src/main/java/io/vertx/core/impl/CloseFuture.java [hunk 3]
```diff
@@ -30,6 +31,7 @@ public class CloseFuture implements Closeable {
   private final Logger log;
   private final Promise<Void> promise;
   private boolean closed;
+  private Map<Closeable, CloseFuture> weakHooks;
   private Map<Closeable, CloseFuture> hooks;
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/core/impl/CloseFuture.java:35]: there is no WeakHashSet afaik

File: src/main/java/io/vertx/core/impl/CloseFuture.java [hunk 4]
```diff
@@ -57,23 +59,37 @@ public synchronized void add(Closeable hook) {
...
-    }
-    if (hooks == null) {
-      hooks = new WeakHashMap<>();
+      fut.future().onComplete(ar -> {
+        remove(fut);
+      });
+      if (weakHooks == null) {
+        weakHooks = new WeakHashMap<>();
+      }
+      weakHooks.put(hook, this);
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/core/impl/CloseFuture.java:68]: it replaces NetClient / HttpClient that can have finalizers

**Comment by vietj** [comment]
@JojOatXGME I wanted you to have a look at the case you provided and check it fixes it

**Comment by JojOatXGME** [comment]
> @JojOatXGME I wanted you to have a look at the case you provided and check it fixes it

Ok. I tested the last SNAPSHOT of Vert.x and I was no-longer able to reproduce the leak. So it seems to be fixed. 👍

**Pull request merged**